ISSUE-404 EXPLAIN should process statements with params placeholders.

kuaw26 opened the pull request.

kuaw26 committed:
ISSUE-404 EXPLAIN should process statements with params placeholders.

svladykin commented:
@kuaw26 Your test fails for remote mode. Please fix it. Otherwise the change looks good to me.

```
ERROR: FAIL org.h2.jdbc.JdbcSQLException: Parameter "#1" is not set [90012-193] org.h2.jdbc.JdbcSQLException: Parameter "#1" is not set [90012-193] ------------------------------
org.h2.jdbc.JdbcSQLException: Parameter "#1" is not set [90012-193]
    at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
    at org.h2.message.DbException.get(DbException.java:179)
    at org.h2.message.DbException.get(DbException.java:155)
    at org.h2.expression.ParameterRemote.checkSet(ParameterRemote.java:48)
    at org.h2.command.CommandRemote.checkParameters(CommandRemote.java:207)
    at org.h2.command.CommandRemote.executeQuery(CommandRemote.java:139)
    at org.h2.jdbc.JdbcStatement.executeQuery(JdbcStatement.java:79)
    at org.h2.test.db.TestCases.checkExplain(TestCases.java:902)
    at org.h2.test.db.TestCases.testExplain(TestCases.java:917)
    at org.h2.test.db.TestCases.test(TestCases.java:78)
    at org.h2.test.TestBase.runTest(TestBase.java:148)
    at org.h2.test.TestAll.addTest(TestAll.java:923)
    at org.h2.test.TestAll.test(TestAll.java:693)
    at org.h2.test.TestAll.runTests(TestAll.java:615)
    at org.h2.test.TestAll.testAll(TestAll.java:537)
    at org.h2.test.TestAll.run(TestAll.java:489)
    at org.h2.test.TestAll.main(TestAll.java:427)
```

kuaw26 commented:
@svladykin,

This test passed locally.
And I do not understand why it failed on Travis.

Could you check my PR locally? And advice how to run it on Travis?

Thanks!

kuaw26 commented:
@svladykin 

Could you give me a hint how to run test in remote mode locally?

svladykin commented:
The easiest way is to initialize `TestAll.networked` field to `true`.

kuaw26 commented:
@svladykin 

Thanks, I reproduced failed test locally.

kuaw26 committed:
ISSUE-404 Fixed CommandRemote to support EXPLAIN with params placeholder.

grandinj commented on h2/src/main/org/h2/command/CommandRemote.java:
```diff
@@ -203,7 +204,12 @@ public int executeUpdate() {
+        boolean explain = isQuery && sql.startsWith("EXPLAIN");
+
         for (ParameterInterface p : parameters) {
+            if (explain)
```
rather check the flag outside the loop

grandinj commented:
looks like TestPreparedStatement is now failing in net=true & memory=true mode, probably that unit test needs tweaking

svladykin commented on h2/src/main/org/h2/command/CommandRemote.java:
```diff
@@ -203,7 +204,12 @@ public int executeUpdate() {
     }
 
     private void checkParameters() {
+        boolean explain = isQuery && sql.startsWith("EXPLAIN");
```
Are you sure this will work for query like `/* bla-bla */ EXPLAIN SELECT ?` ?

grandinj commented:
CommandRemote is an odd place to stick this code. A more logical spot would be in org.h2.command.dml.Explain

kuaw26 commented:
Guys, it seems not trivial to fix CommandRemote correctly.
Query like `/* bla-bla */ EXPLAIN SELECT` are not correctly processed with my current quick fix.
I will investigate how CommandRemote works and try to fix properly.
Do I correctly understand that it serialize SQL and params and pass for remote execution?
And on remote side it will be converted to org.h2.command.dml.Explain?

svladykin commented:
@kuaw26 

Yes, your understanding is correct. 

Overall your approach looks good, but the way you detect `EXPLAIN` inside of `CommandRemote` is obviously too naive.

I think the correct way should be the following. You have to use`CommandInterface.getCommandType()` method, but it is not implemented correctly for `CommandRemote`. In `CommandRemote` constructor we call `prepare(session, true)` which sends everything to the server and receives some meta info back. So this command type must be sent back there by `TcpServerThread`. 

Notice that this will require network protocol change, thus make sure to make it backward compatible by adding new TCP protocol version (like `Constants.TCP_PROTOCOL_VERSION_15`) and checking it where needed.

svladykin commented:
Another thing to notice is that `EXPLAIN ANALYZE` needs the parameters to be checked.

kuaw26 commented:
@svladykin 
Thanks, I will follow your recommendations.

grandinj commented:
@svladykin  that sounds unnecessarily complicated. I'm fairly sure this can be fixed purely server side without any protocol change. Probably by finding an appropriate point inside the Explain class to set the parameters.

svladykin commented:
@grandinj Feel free to propose alternative solution, I'm always for simplicity if possible.

grandinj commented:
@svladykin see my suggestions above, specifically, the org.h2.command.dml.Explain class is a much better place to tweak this sort of stuff - it's already on the server side, so no protocol stuff required, and it works in both remote and embedded mode, and that point we have already parsed the command so we know for sure that we're dealing with an Explain command.

akuznetsov-gridgain commented:
@grandinj

From my (newbe) point of view the main problem is that `CommandRemote` checks params before execution and serialize them for sending to remote execution.

So, I will try as first approach:
1) Remove params check in `CommandRemote` (remote server should do it).
2) Some how serialize null params (may be some sort of stub or special Param).
3) Check everything on remote side.

If this will not work - I will try @svladykin approach.

Thanks.

svladykin commented:
@kuaw26 If you will always put null parameters in `CommandRemote`, then server side checks for all statements will wrongly assume that all the parameters are set.

akuznetsov-gridgain commented:
@svladykin 
Yes, I understand that. And I will try send special param to detect such cases and process them correctly.

svladykin commented:
@kuaw26 This again looks like a compatibility breaking change :)

grandinj commented:
Yeah, I'm sorry, @svladykin is correct. There is currently no nice way to solve this because of the lack of information in the right places.

The correct way would involve bumping the protocol version number and adding a new message SESSION_PREPARE_READ_PARAMS2 which returns the extra information we need (the command type).

grep the code for TCP_PROTOCOL_VERSION to see how we handle versioning compatibility - we want old clients to be able to talk to new servers, and new clients to be able to talk to old servers :-)

kuaw26 commented:
@grandinj, @svladykin 
Thank you! I will investigate.

kuaw26 committed:
ISSUE-404 Fixed CommandRemote to support EXPLAIN with params placeholder.

kuaw26 commented:
@grandinj and @svladykin

Could you, please, take a quick look on my latest changes?
I'm going in right direction?

I'm not finished yet, but I will be very thankful for some tips / critique..

Thanks!

svladykin commented:
I think you are going in right direction, but you have to handle `EXPLAIN ANALYZE` as a special case here, because we need to check parameters for it. Probably we need a separate command type constant `EXPLAIN_ANALYZE`.

Also it looks like something is wrong with tests. May be this is not related to your changes.

kuaw26 committed:
ISSUE-404 Support EXPLAIN with params placeholder WIP.

kuaw26 committed:
ISSUE-404 Support EXPLAIN with params placeholder WIP.

kuaw26 committed:
ISSUE-404 Support EXPLAIN with params placeholder WIP.

kuaw26 committed:
ISSUE-404 Support EXPLAIN with params placeholder WIP.

kuaw26 committed:
ISSUE-404 Muted tests.

kuaw26 commented:
@grandinj and @svladykin
I implemented first version that passed on Travis (except TestKillRestart).
As @grandinj sugested I implemented SESSION_PREPARE_READ_PARAMS2 which returns the extra information.
Also as @svladykin sugested I implemented special support for EXPLAIN ANALYZE (test also added).

Could you please review my changes and give me feedback? May be we need more tests?

svladykin commented on h2/src/main/org/h2/command/CommandRemote.java:
```diff
@@ -204,15 +214,21 @@ public int executeUpdate() {
     private void checkParameters() {
         for (ParameterInterface p : parameters) {
-            p.checkSet();
+            if (cmdType != EXPLAIN)
```
Lets move this condition outside of the loop.

svladykin commented on h2/src/main/org/h2/command/CommandRemote.java:
```diff
@@ -204,15 +214,21 @@ public int executeUpdate() {
+            Value pVal = p.getParamValue();
+
+            if (pVal == null && cmdType == EXPLAIN)
+                pVal = ValueString.get("PARAMETER_NOT_SET");
```
May be just pass `ValueNull.INSTANCE`? Why sending these strings over the wire?

svladykin commented on h2/src/test/org/h2/test/db/TestCases.java:
```diff
@@ -897,6 +898,141 @@ private void testExecuteTrace() throws SQLException {
+    }
+
+    private void testExplain() throws SQLException {
+        deleteDb("cases");
```
Do we have any `EXPLAIN ANALYZE` tests? I do not see any.

svladykin reviewed (changes requested):
Looks good to me, few minor issues to notice though.

kuaw26 committed:
ISSUE-404 Fixed review notes.

akuznetsov-gridgain commented:
@svladykin, thanks for your review!
I fixed notes found on review.
For unknown reason travis failed on openjdk7 and passed on oraclejdk8 (I think it was flaky test).
It is possible restart Travis without pushing commit with fake changes?

kuaw26 committed:
ISSUE-404 Un-muted tests.

kuaw26 committed:
Merge branch 'master' of https://github.com/h2database/h2database into h2-404

kuaw26 commented:
Guys, I merged my PR with latest master.
Test for Java 7 passed.
Could you review my PR and merge it or tell me what I should fix.

grandinj commented:
I'm on holiday, but I'll merge this when I get back in a couple of weeks if @svladykin hasn't done so already. Thanks for all your hard work.

Could we get a license statement to the mailing list please, since this is definitely a significant change :), see here for example:

http://h2database.com/html/build.html#providing_patches

kuaw26 commented:
@grandinj , I added license statement to the H2 Google group: https://groups.google.com/forum/#!topic/h2-database/x86CSZjdfE8

@svladykin could you review my PR once again and give me a feedback or merge it to master.

Thanks!

svladykin commented:
Looks good, thanks a lot for the contribution!
